sdk/src/main/scala/io/horizen/account/state/StateTransition.scala

```diff
@@ -146,4 +143,138 @@ class StateTransition(
     // return remaining gas to the gasPool of the current block so it is available for the next transaction
     blockGasPool.addGas(gas.getGas)
   }
+
+  /**
+   * Execute given invocation on the current call stack.
+   */
+  @throws(classOf[InvalidMessageException])
+  @throws(classOf[ExecutionFailedException])
+  def execute(invocation: Invocation): Array[Byte] = {
+    // limit call depth to 1024
+    if (depth > 1024) throw new ExecutionRevertedException("max call depth exceeded")
+    // get caller gas pool, for the top-level call this is empty
+    // TODO: this will be wrong after a call to `executeDepth()` because it does not add a new invocation to the stack.
```

Comment:
To be honest, I'm not sure because it is not very clear. In my understanding, here it performs a check that the gas passed to the current invocation is not greater than the gas passed to the previous invocation. In fact, in the `invoke` method, that is called by the `execute` method,  the current invocation object is prepended to the `invocationStack` field. So, usually the first element in `invocationStack` is the previous invocation in the call stack. This is not true when the current invocation is a callback from the EVM, nested in more than 1 call of an EVM contract. E.g. 

EVM_Contract.call1 => EVM_Contract.call2=> EVM_Contract.call3=> Native_Contract.callback4

The invocation related to EVM_Contract.call1 is added to `invocationStack` , while the other calls (EVM_Contract.call2,  EVM_Contract.call3) are not, because they are executed directly by the EVM. When Native_Contract.callback4 is executed, `executeDepth()` is called, that calls `execute()` that performs the gas check on the latest Invocation in `invocationStack`, that is related to EVM_Contract.call1. 
So the check may be too optimistic. However, I suppose that the EVM itself will check that the gas passed to a nested call is not too much, so I don't see a real issue. However, I'm not totally sure so I fear I'm missing something...

Reply:
ok! can you add it to the TODO comment?"this will be wrong .... but will not be an issue beacause... "
I want to avoid that someone will read the code in the future and get scared of a possible bug :-)

Reply:
Already done it :)